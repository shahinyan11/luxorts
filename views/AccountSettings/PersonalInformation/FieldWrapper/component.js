import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import classNames from 'classnames';

import GradientButton from 'views/shared/GradientButton';

import useContainer from './hook';

const FieldWrapper = ({
  children,
  label,
  description,
  value,
  hideButtons,
  isLoading,
  changeMode,
  fieldNames,
}) => {
  const { toggleEditMode, editMode, onSave } = useContainer({ changeMode, fieldNames });

  return (
    <div className="personal-information__item mb-24">
      <div className="personal-information__header d-flex justify-content-space-between mb-8">
        <p
          className={classNames('personal-information__label mb-0', {
            'personal-information__label--active': editMode,
          })}
        >
          <FormattedMessage {...label} />
        </p>
        <a
          onClick={toggleEditMode}
          role="button"
          className={classNames('main-text-btn personal-information__edit text-body-2', {
            'personal-information__edit--active': !editMode,
            'personal-information__edit--hidden': editMode,
          })}
        >
          <FormattedMessage id={`shared.${value ? 'edit' : 'add'}`} />
        </a>
      </div>
      {!editMode ? (
        <p
          className={classNames('personal-information__value mb-24', {
            'personal-information__value': value,
            'personal-information__value--disabled': !value,
          })}
        >
          {value || <FormattedMessage id="shared.notProvided" />}
        </p>
      ) : (
        <>
          {description && (
            <p className="personal-information__descr text-body mb-16">
              <FormattedMessage {...description} />
            </p>
          )}
          {children}
          {!hideButtons && (
            <div className="d-flex mb-24">
              <GradientButton
                onClick={toggleEditMode}
                text={{ id: 'shared.cancel' }}
                className="min-width-120 mr-16"
                variant="tertiary"
              />
              <GradientButton
                loading={isLoading}
                onClick={onSave}
                text={{ id: 'shared.save' }}
                className="min-width-140 mr-16"
                variant="primary"
              />
            </div>
          )}
        </>
      )}
    </div>
  );
};

FieldWrapper.defaultProps = {
  description: null,
  value: null,
  hideButtons: false,
  isLoading: false,
  changeMode: true,
  fieldNames: null,
};

FieldWrapper.propTypes = {
  children: PropTypes.node.isRequired,
  label: PropTypes.shape().isRequired,
  fieldNames: PropTypes.arrayOf(PropTypes.string),
  value: PropTypes.string,
  description: PropTypes.shape(),
  hideButtons: PropTypes.bool,
  isLoading: PropTypes.bool,
  changeMode: PropTypes.bool,
};

export default FieldWrapper;
