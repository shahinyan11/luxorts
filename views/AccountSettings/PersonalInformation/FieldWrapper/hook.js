import { useCallback, useState } from 'react';
import { useFormikContext } from 'formik';

import useImmutableCallback from 'hooks/useImmutableCallback';

import isPresent from 'utils/isPresent';
import resetSpecificFields from 'utils/form/resetSpecificFields';

function useContainer({ changeMode, fieldNames }) {
  const formik = useFormikContext();
  const [editMode, setEditMode] = useState(false);

  /**
   * Toggle edit mode
   */
  const toggleEditMode = useCallback(() => {
    resetSpecificFields(formik, fieldNames);

    setEditMode(!editMode);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [editMode]);

  /**
   *  Save updated info and change edit mode
   */
  const onSave = useImmutableCallback(async () => {
    const errors = await formik.validateForm();

    if (!isPresent(errors)) {
      formik.handleSubmit();

      if (changeMode) {
        setEditMode(false);
      }

      return;
    }

    Object.keys(errors).forEach((field) => {
      formik.setFieldTouched(field, true);
    });
  });

  return {
    editMode,
    toggleEditMode,
    onSave,
  };
}

export default useContainer;
