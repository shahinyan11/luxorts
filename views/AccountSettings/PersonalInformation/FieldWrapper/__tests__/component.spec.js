import { shallow } from 'enzyme';

import FieldWrapper from '../component';

const mockedHookData = {
  toggleEditMode: jest.fn(),
  onSave: jest.fn(),
  editMode: false,
};

jest.mock('../hook', () => jest.fn(() => mockedHookData));

describe('FieldWrapper component', () => {
  const props = {
    label: { id: 'shared.field' },
    description: { id: 'personalInformation.field.description' },
    fieldNames: ['firstName', 'lastName'],
  };
  const component = shallow(<FieldWrapper {...props}>Foo</FieldWrapper>);

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
