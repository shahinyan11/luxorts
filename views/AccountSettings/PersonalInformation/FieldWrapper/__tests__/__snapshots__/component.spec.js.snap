// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldWrapper component matches snapshot 1`] = `
<div
  className="personal-information__item mb-24"
>
  <div
    className="personal-information__header d-flex justify-content-space-between mb-8"
  >
    <p
      className="personal-information__label mb-0"
    >
      <Memo(FormattedMessage)
        id="shared.field"
      />
    </p>
    <a
      className="main-text-btn personal-information__edit text-body-2 personal-information__edit--active"
      onClick={[MockFunction]}
      role="button"
    >
      <Memo(FormattedMessage)
        id="shared.add"
      />
    </a>
  </div>
  <p
    className="personal-information__value mb-24 personal-information__value--disabled"
  >
    <Memo(FormattedMessage)
      id="shared.notProvided"
    />
  </p>
</div>
`;
